### DISCUSSION TITLE
Externalize first time config wizard bundled in the InnoSetup Installation

### DISCUSSION THREAD

**Discussion by Sulquendi** [git-for-windows/git#4989] [Ideas]
Hello there !

A bit of context : my company wants to have control over the deployed git versions on our computers, thus leading to build our own git deployment based upon a SCCM Package (InnoSetup /SILENT install). Upgrading the existing versions works fine but we'd like to also use this package for first time installations. So far we have trouble with the innosetup questions to the user (OpenSSL, Credentials Manager, LFS, default editor, ...). Silent installation defaults everything.

Feature proposal : instead of scripting the git config wizard into inno setup, makes a resident stand-alone binary/batch/posh script that would be started in post-config step (or anytime after that). The tool would ask/re-ask the user for the configuration choices.

Thanks !

**Comment by dscho** [comment]
Great idea! I wonder about the best way to go about it, though, without breaking existing users' habits. Suggestions?

**Comment by rimrul** [comment]
> So far we have trouble with the innosetup questions to the user (OpenSSL, Credentials Manager, LFS, default editor, ...). Silent installation defaults everything.

You can override most defaults from the command line. See

https://github.com/git-for-windows/git/wiki/Silent-or-Unattended-Installation
and
https://github.com/git-for-windows/git/wiki/Mapping-between-Git-Installer-GUI-Settings-And-Command-Line-Arguments

**Reply by Sulquendi** [reply]
Thank you for those infos ! 
Theses will be of some use to enforce a config by default. 
Unfortunately, with around 2-3k git users in my company, we cannot set up a config that will fit to everyone, so giving the users the option to set their own config after silent install would still be a major improvement.

**Reply by dscho** [reply]
> giving the users the option to set their own config after silent install would still be a major improvement.

That option already exists: `git config --global [...]`. You can do that for most of the settings (if you know the config setting key). For example, to override the editor used by Git, simply call `git config --global core.editor '<command-line>'` where `<command-line>` might be `code -w` for VS Code.

The biggest advantage of this approach is that it automatically persists through Git for Windows upgrades.

**Reply by Sulquendi** [reply]
> That option already exists: `git config --global [...]`. You can do that for most of the settings (if you know the config setting key). For example, to override the editor used by Git, simply call `git config --global core.editor '<command-line>'` where `<command-line>` might be `code -w` for VS Code.
> 
> The biggest advantage of this approach is that it automatically persists through Git for Windows upgrades.

My idea is to have a GUI to set/get theses options instead of relying only on command line. This GUI already exists during installation. Keeping it as-is seems OK, but having a way to run something similar at any time would be great. I could come up with something but i wonder how to get the setup.iss you are using to build the git setup (or a way to parse any config file to build the gui options with checkboxes and dropdown selects). May be just building the inno setup part that implements the config screens without any files copying/product installation in the registry ?

**Reply by dscho** [reply]
Interesting. I hadn't thought of generating an installer that does not install anything ;-)

As it were, support for this already exists, at least in some form: The canonical way to build an installer is explained [here](https://github.com/git-for-windows/git/wiki/Making-an-installer), and the underlying script [already has an option to skip bundling and installing the files](https://github.com/git-for-windows/build-extra/blob/636e97045adabea061fdcf8bccc3a53b4c882e4a/installer/release.sh#L26-L28). We'd just need to make sure that we don't mistakenly uninstall the previous version, nor that we record a new version having been installed (this latter thing should probably look somewhat like [this](https://github.com/git-for-windows/build-extra/blob/636e97045adabea061fdcf8bccc3a53b4c882e4a/installer/install.iss#L1135)).

Maybe you want to start playing with this idea?
